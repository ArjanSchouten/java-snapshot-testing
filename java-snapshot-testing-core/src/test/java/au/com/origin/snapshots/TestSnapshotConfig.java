package au.com.origin.snapshots;


import java.lang.reflect.Method;
import java.util.stream.Stream;

public class TestSnapshotConfig implements SnapshotConfig {

    @Override
    public StackTraceElement findStacktraceElement() {
        StackTraceElement[] stackTraceElements = Thread.currentThread().getStackTrace();
        int elementsToSkip = 1; // Start after stackTrace
        while (stackTraceElements[elementsToSkip].getClassName()
                .contains(".Snapshot")) {
            elementsToSkip++;
        }

        return Stream.of(stackTraceElements)
                .skip(elementsToSkip)
                .filter(
                        stackTraceElement ->
                                hasTestAnnotation(
                                        ReflectionUtilities.getMethod(
                                                getClassForName(stackTraceElement.getClassName()),
                                                stackTraceElement.getMethodName())))
                .findFirst()
                .orElseThrow(
                        () ->
                                new SnapshotMatchException(
                                        "Could not locate a method with one of supported test annotations"));
    }

    protected boolean hasTestAnnotation(Method method) {
        return method.isAnnotationPresent(org.junit.jupiter.params.ParameterizedTest.class)
                || method.isAnnotationPresent(org.junit.jupiter.api.Test.class)
                || method.isAnnotationPresent(org.junit.jupiter.api.BeforeAll.class);
    }

    private Class<?> getClassForName(String className) {
        try {
            return Class.forName(className);
        } catch (ClassNotFoundException e) {
            throw new IllegalArgumentException(e);
        }
    }
}
