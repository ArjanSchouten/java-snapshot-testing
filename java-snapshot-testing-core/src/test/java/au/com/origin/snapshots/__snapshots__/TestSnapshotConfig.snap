au.com.origin.snapshots.TestSnapshotConfig.findStacktraceElement=[
  {
    "FakeObject.fakeMethod": [
      {
        "arg0": "test1",
        "arg1": 1,
        "arg2": [
          "listTest1"
        ]
      },
      {
        "arg0": "test2",
        "arg1": 2,
        "arg2": [
          "listTest1",
          "listTest2"
        ]
      }
    ]
  }
]


au.com.origin.snapshots.TestSnapshotConfig.findStacktraceElement[Scenario A]=[
  "any type of object"
]


au.com.origin.snapshots.TestSnapshotConfig.findStacktraceElement[Scenario B]=[
  "any second type of object",
  "any third type of object"
]